Queue async operations on a per-player basis [WIP]

sgdc3 opened the pull request.

sgdc3 committed:
Queue async operations on a per-player basis (fixes the order of async tasks)

sgdc3 commented:
Depends on #1992

dependabot-preview[bot] committed:
Bump checker-qual from 3.1.0 to 3.1.1 (#2007)

Bumps [checker-qual](https://github.com/typetools/checker-framework) from 3.1.0 to 3.1.1.
- [Release notes](https://github.com/typetools/checker-framework/releases)
- [Changelog](https://github.com/typetools/checker-framework/blob/master/changelog.txt)
- [Commits](https://github.com/typetools/checker-framework/compare/checker-framework-3.1.0...checker-framework-3.1.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

games647 commented on src/main/java/fr/xephi/authme/process/AsyncUserScheduler.java:
```diff
@@ -0,0 +1,92 @@
+        }
+
+        public void submit(K key, Runnable runnable) {
+            executors.computeIfAbsent(key, k -> new SequentialExecutor(scheduler, () -> executors.remove(key)))
```
Potential memory leak

games647 commented on src/main/java/fr/xephi/authme/process/AsyncUserScheduler.java:
```diff
@@ -0,0 +1,92 @@
+        }
+
+        public void submit(Runnable task) {
+            queue.add(task);
```
This doesn't seem thread-safe. Considering two calls to the `submit` method. Then both tasks will be added, but `scheduler.apply()` will be invoked twice with two new tasks on the Bukkit scheduler resulting to two new threads. 

T1 and T2 are the threads:
T1: 1 iteration of the while loop
-- 1 task remaining
T1: evaluate `while(!queue.isEmpty())` -> `true` - enter `while` loop
T2: 1 iteration of the while loop
-- `queue.poll` wasn't executed yet by T1 and so `queue` is still not empty
-- T2 now removes the last task
T1: `queue.poll()` - `NoSuchElementExeception`

sgdc3 replied:
Hmm I see, which pattern should we use instead?

games647 replied:
The easiest solution would be to lock on the queue on `isEmpty` + `poll` operations.

TuxCoding reviewed (changes requested):
I kind of understand what you are doing here. You are trying to perform asynchronous tasks sequential on per player basis. I guess with the goal to reduce conflicts between them? Do you also thought about locks? This allows more flexibility like read/write differentiation, protecting only thread-unsafe operations meanwhile other parts could execute in `concurrently` and in `parallel`.

sgdc3 commented:
@games647 i think that both read/write operations needs to be sequential, I can't think about any situation where locks might be better. Can you make some examples?

dependabot-preview[bot] committed:
Bump postgresql from 42.2.9 to 42.2.10 (#2010)

Bumps [postgresql](https://github.com/pgjdbc/pgjdbc) from 42.2.9 to 42.2.10.
- [Release notes](https://github.com/pgjdbc/pgjdbc/releases)
- [Changelog](https://github.com/pgjdbc/pgjdbc/blob/master/CHANGELOG.md)
- [Commits](https://github.com/pgjdbc/pgjdbc/compare/REL42.2.9...REL42.2.10)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

ljacqu committed:
#1448 Create AuthMePlayer to get player data from API with (#2000)

* #1448 Create AuthMePlayer to get player data from API with

* #1448 Add tests for new API method & AuthMePlayer

* #1448 Create AuthMePlayer to get player data from API with
- Use Optional for all values that may be null

* #1448 Add comment that AuthMePlayer data does not update itself

dependabot-preview[bot] committed:
Bump maven-shade-plugin from 3.2.1 to 3.2.2 (#2016)

Bumps [maven-shade-plugin](https://github.com/apache/maven-shade-plugin) from 3.2.1 to 3.2.2.
- [Release notes](https://github.com/apache/maven-shade-plugin/releases)
- [Commits](https://github.com/apache/maven-shade-plugin/compare/maven-shade-plugin-3.2.1...maven-shade-plugin-3.2.2)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

sgdc3 committed:
Pom cleanup

Addresses #2014

dependabot-preview[bot] committed:
Bump mockito-core from 3.2.4 to 3.3.0

Bumps [mockito-core](https://github.com/mockito/mockito) from 3.2.4 to 3.3.0.
- [Release notes](https://github.com/mockito/mockito/releases)
- [Commits](https://github.com/mockito/mockito/compare/v3.2.4...v3.3.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

ljacqu committed:
Fix failing tests after Mockito upgrade

dependabot-preview[bot] committed:
Bump maxmind-db-gson from 2.0.2-SNAPSHOT to 2.0.3 (#2034)

Bumps [maxmind-db-gson](https://github.com/maxmind/MaxMind-DB-Reader-java) from 2.0.2-SNAPSHOT to 2.0.3.
- [Release notes](https://github.com/maxmind/MaxMind-DB-Reader-java/releases)
- [Changelog](https://github.com/maxmind/MaxMind-DB-Reader-java/blob/master/CHANGELOG.md)
- [Commits](https://github.com/maxmind/MaxMind-DB-Reader-java/commits)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
Bump checker-qual from 3.1.1 to 3.2.0 (#2033)

Bumps [checker-qual](https://github.com/typetools/checker-framework) from 3.1.1 to 3.2.0.
- [Release notes](https://github.com/typetools/checker-framework/releases)
- [Changelog](https://github.com/typetools/checker-framework/blob/master/changelog.txt)
- [Commits](https://github.com/typetools/checker-framework/compare/checker-framework-3.1.1...checker-framework-3.2.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
Bump maven-site-plugin from 3.8.2 to 3.9.0 (#2041)

Bumps [maven-site-plugin](https://github.com/apache/maven-site-plugin) from 3.8.2 to 3.9.0.
- [Release notes](https://github.com/apache/maven-site-plugin/releases)
- [Commits](https://github.com/apache/maven-site-plugin/compare/maven-site-plugin-3.8.2...maven-site-plugin-3.9.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
Bump postgresql from 42.2.10 to 42.2.11 (#2040)

Bumps [postgresql](https://github.com/pgjdbc/pgjdbc) from 42.2.10 to 42.2.11.
- [Release notes](https://github.com/pgjdbc/pgjdbc/releases)
- [Changelog](https://github.com/pgjdbc/pgjdbc/blob/master/CHANGELOG.md)
- [Commits](https://github.com/pgjdbc/pgjdbc/compare/REL42.2.10...REL42.2.11)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

Twonox committed:
Update messages_fr.yml (#2035)

Legion64 committed:
Update messages_tr.yml (#2052)

Updated Turkish Lang

1337god committed:
Update messages_pt.yml (#2047)

ljacqu committed:
Update checkstyle config
- "allowMissingThrowsTags" does nothing anymore, cf. https://github.com/checkstyle/checkstyle/issues/7329

dependabot-preview[bot] committed:
Bump googleauth from 1.4.0 to 1.5.0 (#2062)

Bumps [googleauth](https://github.com/wstrange/GoogleAuth) from 1.4.0 to 1.5.0.
- [Release notes](https://github.com/wstrange/GoogleAuth/releases)
- [Changelog](https://github.com/wstrange/GoogleAuth/blob/master/NEWS)
- [Commits](https://github.com/wstrange/GoogleAuth/compare/1.4.0...1.5.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
Bump checker-qual from 3.2.0 to 3.3.0 (#2058)

Bumps [checker-qual](https://github.com/typetools/checker-framework) from 3.2.0 to 3.3.0.
- [Release notes](https://github.com/typetools/checker-framework/releases)
- [Changelog](https://github.com/typetools/checker-framework/blob/master/changelog.txt)
- [Commits](https://github.com/typetools/checker-framework/compare/checker-framework-3.2.0...checker-framework-3.3.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

sgdc3 committed:
Merge branch 'master' into player-task-queue

sgdc3 committed:
Merge branch 'master' into player-task-queue

dependabot-preview[bot] committed:
Bump maven-javadoc-plugin from 3.1.1 to 3.2.0 (#2044)

Bumps [maven-javadoc-plugin](https://github.com/apache/maven-javadoc-plugin) from 3.1.1 to 3.2.0.
- [Release notes](https://github.com/apache/maven-javadoc-plugin/releases)
- [Commits](https://github.com/apache/maven-javadoc-plugin/compare/maven-javadoc-plugin-3.1.1...maven-javadoc-plugin-3.2.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
Bump mockito-core from 3.3.0 to 3.3.3 (#2045)

Bumps [mockito-core](https://github.com/mockito/mockito) from 3.3.0 to 3.3.3.
- [Release notes](https://github.com/mockito/mockito/releases)
- [Commits](https://github.com/mockito/mockito/compare/v3.3.0...v3.3.3)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
Bump postgresql from 42.2.11 to 42.2.12 (#2056)

Bumps [postgresql](https://github.com/pgjdbc/pgjdbc) from 42.2.11 to 42.2.12.
- [Release notes](https://github.com/pgjdbc/pgjdbc/releases)
- [Changelog](https://github.com/pgjdbc/pgjdbc/blob/master/CHANGELOG.md)
- [Commits](https://github.com/pgjdbc/pgjdbc/compare/REL42.2.11...REL42.2.12)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
Bump maven-shade-plugin from 3.2.2 to 3.2.3 (#2076)

Bumps [maven-shade-plugin](https://github.com/apache/maven-shade-plugin) from 3.2.2 to 3.2.3.
- [Release notes](https://github.com/apache/maven-shade-plugin/releases)
- [Commits](https://github.com/apache/maven-shade-plugin/compare/maven-shade-plugin-3.2.2...maven-shade-plugin-3.2.3)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
Bump argon2-jvm-nolibs from 2.6 to 2.7 (#2077)

Bumps [argon2-jvm-nolibs](https://github.com/phxql/argon2-jvm) from 2.6 to 2.7.
- [Release notes](https://github.com/phxql/argon2-jvm/releases)
- [Changelog](https://github.com/phxql/argon2-jvm/blob/master/CHANGELOG.md)
- [Commits](https://github.com/phxql/argon2-jvm/compare/v2.6...v2.7)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

DevLeoko committed:
Fixed unrestricted inventories (#2080)

Commit:
Update messages_ru.yml (#2079)

* Update messages_ru.yml

* Update messages_ru.yml

ljacqu committed:
Fix failing tests

dependabot-preview[bot] committed:
Bump HikariCP from 3.4.2 to 3.4.4 (#2101)

Bumps [HikariCP](https://github.com/brettwooldridge/HikariCP) from 3.4.2 to 3.4.4.
- [Release notes](https://github.com/brettwooldridge/HikariCP/releases)
- [Changelog](https://github.com/brettwooldridge/HikariCP/blob/dev/CHANGES)
- [Commits](https://github.com/brettwooldridge/HikariCP/compare/HikariCP-3.4.2...HikariCP-3.4.4)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
Bump sqlite-jdbc from 3.30.1 to 3.31.1 (#2100)

Bumps [sqlite-jdbc](https://github.com/xerial/sqlite-jdbc) from 3.30.1 to 3.31.1.
- [Release notes](https://github.com/xerial/sqlite-jdbc/releases)
- [Changelog](https://github.com/xerial/sqlite-jdbc/blob/master/CHANGELOG)
- [Commits](https://github.com/xerial/sqlite-jdbc/compare/3.30.1...3.31.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
Bump checker-qual from 3.3.0 to 3.4.0 (#2097)

Bumps [checker-qual](https://github.com/typetools/checker-framework) from 3.3.0 to 3.4.0.
- [Release notes](https://github.com/typetools/checker-framework/releases)
- [Changelog](https://github.com/typetools/checker-framework/blob/master/changelog.txt)
- [Commits](https://github.com/typetools/checker-framework/compare/checker-framework-3.3.0...checker-framework-3.4.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
Bump HikariCP from 3.4.4 to 3.4.5 (#2107)

Bumps [HikariCP](https://github.com/brettwooldridge/HikariCP) from 3.4.4 to 3.4.5.
- [Release notes](https://github.com/brettwooldridge/HikariCP/releases)
- [Changelog](https://github.com/brettwooldridge/HikariCP/blob/dev/CHANGES)
- [Commits](https://github.com/brettwooldridge/HikariCP/compare/HikariCP-3.4.4...HikariCP-3.4.5)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
Bump api from 5.0 to 5.1 (#2106)

Bumps [api](https://github.com/lucko/LuckPerms) from 5.0 to 5.1.
- [Release notes](https://github.com/lucko/LuckPerms/releases)
- [Commits](https://github.com/lucko/LuckPerms/compare/v5.0...v5.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

Madis0 committed:
Estonian improvements (#2081)

Yudi committed:
Update messages_br.yml (#2086)

* Update messages_br.yml

* Update messages_br.yml

sgdc3 committed:
Update docs

sgdc3 committed:
Bump commons-email version

Should fix issues due to old Email API usages

dependabot-preview[bot] committed:
Bump checker-qual from 3.4.0 to 3.4.1 (#2123)

Bumps [checker-qual](https://github.com/typetools/checker-framework) from 3.4.0 to 3.4.1.
- [Release notes](https://github.com/typetools/checker-framework/releases)
- [Changelog](https://github.com/typetools/checker-framework/blob/master/changelog.txt)
- [Commits](https://github.com/typetools/checker-framework/compare/checker-framework-3.4.0...checker-framework-3.4.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
Bump maven-shade-plugin from 3.2.3 to 3.2.4 (#2124)

Bumps [maven-shade-plugin](https://github.com/apache/maven-shade-plugin) from 3.2.3 to 3.2.4.
- [Release notes](https://github.com/apache/maven-shade-plugin/releases)
- [Commits](https://github.com/apache/maven-shade-plugin/compare/maven-shade-plugin-3.2.3...maven-shade-plugin-3.2.4)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
Bump postgresql from 42.2.12 to 42.2.14 (#2131)

Bumps [postgresql](https://github.com/pgjdbc/pgjdbc) from 42.2.12 to 42.2.14.
- [Release notes](https://github.com/pgjdbc/pgjdbc/releases)
- [Changelog](https://github.com/pgjdbc/pgjdbc/blob/master/CHANGELOG.md)
- [Commits](https://github.com/pgjdbc/pgjdbc/compare/REL42.2.12...REL42.2.14)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
Bump sqlite-jdbc from 3.31.1 to 3.32.3 (#2136)

Bumps [sqlite-jdbc](https://github.com/xerial/sqlite-jdbc) from 3.31.1 to 3.32.3.
- [Release notes](https://github.com/xerial/sqlite-jdbc/releases)
- [Changelog](https://github.com/xerial/sqlite-jdbc/blob/master/CHANGELOG)
- [Commits](https://github.com/xerial/sqlite-jdbc/compare/3.31.1...3.32.3)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

dependabot-preview[bot] committed:
Bump maven-site-plugin from 3.9.0 to 3.9.1 (#2139)

Bumps [maven-site-plugin](https://github.com/apache/maven-site-plugin) from 3.9.0 to 3.9.1.
- [Release notes](https://github.com/apache/maven-site-plugin/releases)
- [Commits](https://github.com/apache/maven-site-plugin/compare/maven-site-plugin-3.9.0...maven-site-plugin-3.9.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

galaipa committed:
Update Euskara (basque) translation (#2138)

sgdc3 committed:
Merge branch 'master' into player-task-queue
